### PR TITLE
[RFC] API Versioning

### PR DESCRIPTION
**Pull request by ksuderman**
Addresses #242 

I will delete this PR and submit a proper one shortly,  but I wanted a quick review by @almahmoud to verify this is what you have in mind.  Once I am able to bend Kubernetes and Helm to my will I can do some proper testing under the various Kubernetes versions (or at least I think I will...)

### CODE REVIEW

**Review by almahmoud — APPROVED** [review]
This looks great!

File: galaxy/templates/ingress-activity-canary.yaml [hunk 1]
```diff
@@ -7,10 +7,13 @@
...
+apiVersion: networking.k8s.io/v1
+{{- else if semverCompare "^1.14" $k8s_version -}}
 apiVersion: networking.k8s.io/v1beta1
-{{- else -}}
-apiVersion: extensions/v1beta1
+{{ else }}
+{{ fail "This chart requires Kubernetes v1.14 or later" }}
 {{- end }}
 kind: Ingress
 metadata:
```
Comment (almahmoud) [inline galaxy/templates/ingress-activity-canary.yaml:19]: Need to change the paths form below for the new API version:
```
            backend:
              service:
                name: {{ $fullName }}-nginx
                port:
                  number: {{ $servicePort }}
```

roughly

File: galaxy/templates/ingress.yaml [hunk 2]
```diff
@@ -1,10 +1,14 @@
...
+apiVersion: networking.k8s.io/v1
+{{- else if semverCompare "^1.14" $k8s_version -}}
 apiVersion: networking.k8s.io/v1beta1
-{{- else -}}
-apiVersion: extensions/v1beta1
+{{ else }}
+{{ fail "This chart requires Kubernetes v1.14 or later" }}
 {{- end }}
 kind: Ingress
 metadata:
```
Comment (almahmoud) [inline galaxy/templates/ingress.yaml:14]: Similar to above comment, need to change the paths form below.

```
            backend:
              service:
                name: {{ $fullName }}
                port:
                  number: {{ $svcPort }}
```
roughly

**Review by almahmoud — CHANGES_REQUESTED** [review]
I resolved conflicts with the latest version given the helm3 PR got merged. Just a few changes and this should be good for merging.

File: galaxy/templates/ingress.yaml [hunk 3]
```diff
@@ -35,7 +39,7 @@ spec:
           - path: {{ .path }}
             backend:
               serviceName: {{ $fullName }}
-              servicePort: {{ $svcPort }}
+              servicePort: {{ $servicePort }}
```
Comment (almahmoud) [inline galaxy/templates/ingress.yaml:42]:
```suggestion
              service:
                name: {{ $fullName }}
                port:
                  number: {{ $servicePort }}
```

File: galaxy/values.yaml [hunk 4]
```diff
@@ -97,7 +97,7 @@ persistence:
   enabled: true
   name: galaxy-pvc
   # annotations: {}
-  # storageClass:
+  storageClass: nfs
```
Comment (almahmoud) [inline galaxy/values.yaml:100]: We don't want a default storageClass so that it uses the cluster default if not set.

File: galaxy/values.yaml [hunk 5]
```diff
@@ -180,7 +180,7 @@ postgresql:
   existingDatabase:
   galaxyDatabaseUser: galaxydbuser
   # Password is autogenerated and stored in a secret if not specified
-  # galaxyDatabasePassword:
+  galaxyDatabasePassword: galaxydbpassword
```
Comment (almahmoud) [inline galaxy/values.yaml:183]: We shouldn't hardcode the password. It's convenient for dev but not a good default overall

**Review by almahmoud — CHANGES_REQUESTED** [review]

File: galaxy/values.yaml [hunk 6]
```diff
@@ -7,7 +7,7 @@ fullnameOverride: ""
 
 image:
   repository: galaxy/galaxy-k8s
-  tag: 20.09
+  tag: 21.01
```
Comment (almahmoud) [inline galaxy/values.yaml:10]: We're probably just waiting for 4.0.0 to change default to `21.01`

**Comment by nuwang** [comment]
It doesn't quite work for me locally. Could it be the dry run?
```
~/work/galaxy-helm/galaxy(ksuderman-242-api-versioning ✗) kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2-rc2+k3s1", GitCommit:"d38505b124c92bffd45f6e0654adb9371cae9610", GitTreeState:"clean", BuildDate:"2020-09-18T17:34:53Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}

~/work/galaxy-helm/galaxy(ksuderman-242-api-versioning ✗) helm version
version.BuildInfo{Version:"v3.5.3", GitCommit:"041ce5a2c17a58be0fcd5f5e16fb3e7e95fea622", GitTreeState:"dirty", GoVersion:"go1.15.8"}

~/work/galaxy-helm/galaxy(ksuderman-242-api-versioning ✗) helm install --dry-run --debug test .
install.go:173: [debug] Original chart version: ""
install.go:190: [debug] CHART PATH: /Users/nuwan/work/galaxy-helm/galaxy

Error: template: galaxy/templates/priorityclass-job.yaml:8:5: executing "galaxy/templates/priorityclass-job.yaml" at <fail "This chart requires Kubernetes v1.14 or later">: error calling fail: This chart requires Kubernetes v1.14 or later
helm.go:81: [debug] template: galaxy/templates/priorityclass-job.yaml:8:5: executing "galaxy/templates/priorityclass-job.yaml" at <fail "This chart requires Kubernetes v1.14 or later">: error calling fail: This chart requires Kubernetes v1.14 or later
```

**Comment by nuwang** [comment]
Looks like that's not the reason either:
```
~/work/galaxy-helm/galaxy(ksuderman-242-api-versioning ✗) helm install  test .
Error: template: galaxy/templates/priorityclass-job.yaml:8:5: executing "galaxy/templates/priorityclass-job.yaml" at <fail "This chart requires Kubernetes v1.14 or later">: error calling fail: This chart requires Kubernetes v1.14 or later
```

**Comment by nuwang** [comment]
Thanks @ksuderman This fixes the deployments and works as expected.

**Pull request merged**